### DISCUSSION TITLE
Usage of getters and setters for virtual and calculates attributes

### DISCUSSION THREAD

**Discussion by zirkelc** [tywalch/electrodb#383] [Q&A] [answered]
Hi,

I'm trying to use getters, setters and watch to create a combination of calculated an virtual attributes. I have the three attributes  `contextId`, `contextType` and `contextUrl` that I want to have as individual primitive attributes to be able to use them as secondary index. 

However, when reading an entity I'd like to merge these three attributes into one object attribute `context` and hide the three individual attributes from the result. When setting the context, ideally, I'd like to set the `context` attribute and it distributes the values to the three attributes (or they react to the change of `context` via watch). The `context` attribute should only exist on the result, but not be persisted to the database.

Example:

```ts
await tasks
  .patch({ id: "1" })
  // this should set the values on contexId, contextType and contextUrl
  .set({ context: { id: "1", type: "product", url: "https://example.com" } })
  .go();

// result should contain context without contexId, contextType and contextUrl
const result = tasks.get({ id }).go();
```

[Playground](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAY1IBfbhiiSwAIkRwjVSzABGNgNzijUPrgVUGFwRSAF5bTShUDAB9AKC4aL4GSTgPcVAwABUGXABrITUNKk1hMW9ffxz8sDC+OBZVdS0AClEwJTb2sElYRAAuDq6u-iLNAZsAvNwbfE6h-UxcLB9xgEYZua7iDEMTccnchmlpG03TWaGGKicsF1QqPAHFTfasGCeXruLpOHG-bD4AHMNkMhuQ0FhyO8wE50J9zp9yg8AB5UACS0OeoK+mh+fxuQJB2LAdDe8D4A1hcAuxJYVyMdAGAG0bEi4KibABdGnY4hUAbNaKEZRs1FgUwAShqAD5BsT2uQqFE+GBRVQAHRvTzy0zwnldNWZXG-OXY74mmz-LCE-VDUkwcmUjDoW1dOlUBnM1k+FFULmu9p8gVCpSqn3shSSmWm4mK5Vh9QR9Xm7XE3XYhHYtUAVQwCA+8vN+IBwIDJLJ-CdLs+7XdnrALLV-prRDg-LAguFCd94qlIVlWPlZDb8bV6qiCFTGb1iPDqILxKLYAA5JIjsuy9JLD95DhcAv5ROnjDjQNl1agcvxWX2kvlEvzwTAVfM0O3seHxfn9eW6-sYC22DLs1QxQhDWNMC5yoXMkCjfsY2xOMMBVQchzAd9uwjUCWyGJdwJ+G8uiPTDURgnDxSnUE-1BYBgFbdtmj7WVUAEOAuGtOAYFdajqOteBkUeBCwC3KAACsHBoA9QWkXIpOxdjEGhGwZKJeVvBkKAlgeL03i5X8yzyOTQQUhAlLyVTiXU2QtJNJlOV-GchnTH92nvQJgnFUQJU8UQMgAdWwB4UHQXghCoKAwAAuoMCuUhpCYZI20WMAfBhOhSGwQE6AUEQvB8Px0JgGpbHWHyDlwTp1Xij06GaZQ3l7TpaPojQgTS0g1XoSJTPo9qwH0BgEHQIRUtFUCSKoI0fjABgBAmsj2nVPk6om48MJsdZCCXZTnFQRwZjAYibGyqhpH3WibHFRrFsBKBGLKqpcHVACqBWhrJWeu6JWTdK+GachcEoBR4MHAGgYAfnVGArgYdU1U8SUgA)

Currently, `context` is persisted with the three attributes. I think to not persist `context` I would have to use `set: () => undefined` but then the other attributes would never receives an update. Maybe it's simply not possible to have this kind of dependency between attributes?

**Comment by tywalch** [comment, marked as answer]
Hi @zirkelc 

Ultimately `watch` is more to help with some basic/common denormalization patterns, and less to perform object transformation. For those types of transformations, which I think this is one of those, it's ultimately left to the user to create the end-state objects they need. ElectroDB acts maybe less like an ORM and more like query builder in this regard.

**Reply by zirkelc** [reply]
Okay, that's what I thought. I just wanted to make sure I hadn't overlooked anything and that this is actually possible. 

Thanks for the quick reply!

**Reply by tywalch** [reply]
One thing I'll add, to help you reason about why you did see what you had hoped in your case: Any attributes "watched" by another attribute will have their setters called prior to calling the setter of the "watcher". So in your case you'd lack a mechanism to prevent writing the context object
